serdes/serde-common/src/main/java/io/apicurio/registry/serde/K4ERCache.java

```diff
@@ -0,0 +1,295 @@
+
+    private void checkInitialized() {
+        boolean initialized = keyExtractor1 != null && keyExtractor2 != null &&
+            keyExtractor3 != null && keyExtractor4 != null;
+        initialized = initialized && (loaderFunction1 != null || loaderFunction2 != null ||
+            loaderFunction3 != null || loaderFunction4 != null);
+        initialized = initialized && lifetime != null && backoff != null && retries >= 0;
+        if (!initialized)
+            throw new IllegalStateException("Not properly initialized!");
+    }
+
+    // === Specific
+
+    public V get1(K1 key) {
+        return get1(key, null);
+    }
+
+    public V get1(K1 key, Object context) {
+        Long id = this.index1.get(key);
+        // If id == null, we don't have the value in the cache,
+        // because all key extractors have to work.
+        if (id == null)
+            id = this.lastId.incrementAndGet();
+        return getValue(id, key, loaderFunction1, context);
+    }
+
+    public V get2(K2 key) {
+        return get2(key, null);
+    }
+
+    public V get2(K2 key, Object context) {
+        Long id = this.index2.get(key);
+        // If id == null, we don't have the value in the cache,
+        // because all key extractors have to work.
+        if (id == null)
+            id = this.lastId.incrementAndGet();
+        return getValue(id, key, loaderFunction2, context);
+    }
+
+    public V get3(K3 key) {
+        return get3(key, null);
+    }
+
+    public V get3(K3 key, Object context) {
+        Long id = this.index3.get(key);
+        // If id == null, we don't have the value in the cache,
+        // because all key extractors have to work.
+        if (id == null)
+            id = this.lastId.incrementAndGet();
+        return getValue(id, key, loaderFunction3, context);
+    }
+
+    public V get4(K4 key) {
+        return get4(key, null);
+    }
+
+    public V get4(K4 key, Object context) {
+        Long id = this.index4.get(key);
+        // If id == null, we don't have the value in the cache,
+        // because all key extractors have to work.
+        if (id == null)
+            id = this.lastId.incrementAndGet();
+        return getValue(id, key, loaderFunction4, context);
+    }
+
+    // === Generic
+
+    private <T> V getValue(Long id, T key, BiFunction<T, Object, V> loaderFunction, Object context) {
+        checkInitialized();
+        WrappedValue<V> value = cache.compute(id, (id0, wrappedValue) -> {
+            if (wrappedValue == null) {
+                // With retry
+                Optional<V> newValue = retry(backoff, retries, () -> {
+                    return loaderFunction.apply(key, context);
+                });
+                if (newValue.isPresent()) {
+                    // Index
+                    reindex(id, newValue.get());
+                    // Return
+                    return new WrappedValue<>(lifetime, Instant.now(), newValue.get());
+                }
+                return null;
```

Comment:
this breaks compatibility. The schema resolver cannot return null values, if there is no way to get a shema to return it has to throw an exception. That's how it's working currently, and changing that will make all of our serdes to now throw `NullPointerException`